### ISSUE TITLE
Save Consult for Public Body

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#3568]
* As an IAO analyst
* I want to be able to save a consult without redactions
* So that I can consult with a public body(s) on pages in an FOI request

**Assumptions & Scope**	
What are the assumptions for this story?

- For release 8.0, the ministry will not have access to the Records Review app.
-  The request must be in the Records Review state.
- Every page in the request package must have at least one of these flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive, Duplicate or Non Responsive 
- Packages with pages with no flag or the 'Consult' or 'In Progress' flag cannot be saved as a redline
- All redactions will be red boxes with the content visible.
- All records will be saved in one PDF document in Object Storage.
- And all Duplicate and Not Responsive flagged pages will be removed from the redline PDF
- The PDF document will include all redactions, applied  with a red box around the visible content, with the section stamp.
- The PDF document will include NO comments, highlights or annotations. 
- The PDF document will be viewable in the redaction app.

What is IN scope?
Saving the redline in Object Storage.

Create Records Package dropdown

What is NOT in scope?
Creating redactions.

State changes.

Saved phased releases of the redline will be in future story.

A dropdown to toggle between layers: 'Records Package', Redline, IAO Review, Harms, Ministry Sign Off

Creating a new records package.

**Acceptance Criteria**	

**Scenario 1: Active "Create Records Package" Dropdown**
* GIVEN an analyst is on a request
* WHEN IAO analyst is in the redaction app viewing the records
* THEN there will be an active "Create Records Package" dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460

**Scenario 2: Click on "Create Records Package" Dropdown**
* GIVEN an IAO analyst is in the redaction app viewing the records
* WHEN IAO analyst clicks on the "Create Records Package" Dropdown
* THEN the dropdown will open
* AND there will 3 options: 'Consult Public Bodies,' 'Redline for Sign Off,' 'Final Package for Applicant.'

**Scenario 3: Inactive "Redline for Sign Off"**
* GIVEN a records package is not in Records Review or Ministry Sign Off
* AND every page in the Records Package has one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst is clicks on the "Create Records Package" Dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* THEN there will be inactive and greyed out "Redline for Sign Off" option in the dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460

**Scenario 4: Inactive 'Redline for Sign Off"**
* GIVEN a records package is in Records Review or Ministry Sign Off
* AND every page in the Records Package does not have one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst is clicks on the "Create Records Package" Dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* THEN there will be inactive and greyed out "Redline for Sign Off" option in the dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460

**Scenario 5: Active 'Redline for Sign Off'**
* GIVEN a records package is in Records Review or Ministry Sign Off
* AND every page in the Records Package has one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst clicks on the Create Records Package" dropdown
* THEN the "Redline for Sign Off" will be clickable
* AND the "Final Package for Applicant" will not be clickable
* AND the "Consult Public Body" will not be clickable (i.e. no pages are flagged with Consult)

**Scenario 6: Click "Redline for Sign Off"**
* GIVEN the Create Records Package" dropdown is open
* AND the request is in Records Review
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* WHEN the IAO analyst clicks the active "Redline for Sign Off"
* THEN a modal will open with the message, "Are you sure want to create the redline for ministry sign off. 
* AND a "Create Redline" and "Cancel" button
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459

**Scenario 7: Click "Cancel" button**
* GIVEN an analyst is on the "Create Redline" modal
* WHEN the IAO analyst clicks the "Cancel" button
* THEN a modal will close
* AND the analyst will remain in the redaction app
* AND no redline PDF will be saved

**Scenario 8: Click "Create Redline" button**
* GIVEN an analyst is on the "Create Redline" modal
* WHEN the IAO analyst clicks the "Create Redline" button
* THEN a modal will close
* AND a green success toast will appear when the save is complete
* AND the analyst will remain in the redaction app
* AND the analyst will see the created redline PDF in the redaction app so they can review if needed
* AND one redline PDF with all records will be saved in Object Storage
* AND all pages flagged as Duplicate or Non Responsive pages will be removed from the redline
* AND the redline PDF will contain all redactions (red boxes around visible text with the section of the ACT stamped in the box)
* AND the redline PDF will contain no comments or annotations of any kind

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by lmullane** [comment]
@m-prodan, @KyEggleston , @liseandtea, @JHarrietha-AOT 

Please review and comment. 

@JHarrietha-AOT, demo feedback is that analyst need a way to save pages for consult (this story) and export/download the pages (a story to come). Design updates will be needed.

**Comment by KyEggleston** [comment]
The only question that comes to mind is how we will handle when a user adds redactions to Consult PDF(s) (after they have been created through 'Create Consult'). Will they need to select 'Create Consult' again to save their redactions? Or will we have a separate save button (or are the updates automatically saved)?

Just thinking about how this would work with our 'Download for Consult' option (to download the Consult PDFs). Scenario 6 in #3570 mentions:

"WHEN there have been changes made to the Consult, including redactions
OR if pages have been added or removed
**_AND the changes have been saved_**"

How will saving work?

**Comment by m-prodan** [comment]
Assumption for scenario 5 is that we'll need a design element to show where the user can switch to the consult records.

Other feedback - scenario 2, we may need to make sure we revisit how those disabled options need to become enabled. I am thinking specifically things like phased releases. A phased release may also happen if we are waiting on a consult (ie: we've got most of the records reviewed/ready for release, except for outstanding consults) 

Otherwise - looks good, will continue reviewing the other stories

**Comment by lmullane** [comment]
@abin-aot, is this still 40 points given that we are now stitching the documents for redaction?

cc @m-prodan
